src/app.jsx

```diff
@@ -44,23 +50,101 @@ export default class App extends React.Component {
     this.handleFileSelect = this.handleFileSelect.bind(this);
     this.onSelectData = this.onSelectData.bind(this);
     this.onSelectColorPack = this.onSelectColorPack.bind(this);
+    this.toggleUserData = this.toggleUserData.bind(this);
+    this.editHeader = this.editHeader.bind(this);
+    this.saveEditedHeaders = this.saveEditedHeaders.bind(this);
+    this.addHeaders = this.addHeaders.bind(this);
+    this.showHeaderDialog = this.showHeaderDialog.bind(this);
+    this.cancelEditDialog = this.cancelEditDialog.bind(this);
   }
 
   handleFileSelect(evt) {
     const file = evt.target.files[0];
 
     Papa.parse(file, {
       complete: (results) => {
-        this.setState({data: results.data});
+        this.setState({data: results.data, headerRow: results.data[0]});
       },
     });
   }
 
+  toggleUserData(evt) {
+    if (evt.target.value === 'true') {
+      this.setState({useSampleData: true, data: data.smallData});
+    } else {
+      this.setState({useSampleData: false, data: [[]]});
+    }
+  }
+
+  showHeaderDialog(e) {
+    const editType = e.target.value;
+
+    if (editType === 'editHeaders') {
+      this.setState({showEditHeaderForm: true, showAddHeaderForm: false,
+        headerRow: this.state.data[0]});
+    } else if (editType === 'addHeaders') {
+      const blankHeaderRow = [];
+
+      this.state.data[0].forEach(header => {
+        blankHeaderRow.push(null);
+      });
+      this.setState({showAddHeaderForm: true, showEditHeaderForm: false,
+        headerRow: blankHeaderRow});
+    }
+  }
+
+  cancelEditDialog() {
+    this.setState({showEditHeaderForm: false, showAddHeaderForm: false,
+      headerRow: this.state.data[0], validationMessage: null});
+  }
+
+  saveEditedHeaders() {
+    const newDataSet = this.state.data.slice();
+
+    newDataSet[0] = this.state.headerRow;
+    this.setState({data: newDataSet, showEditHeaderForm: false});
+  }
+
+  addHeaders() {
+    const newDataSet = this.state.data.slice();
+
+    let validHeaders = true;
+
+    this.state.headerRow.forEach(header => {
+      if (header === null || header === '') {
+        validHeaders = false;
+      }
+    });
+    if (validHeaders === false) {
+      this.setState({validationMessage: 'Headers cannot be blank.'});
+    } else {
+      newDataSet.unshift(this.state.headerRow);
+      this.setState({data: newDataSet, showAddHeaderForm: false,
+        validationMessage: null});
+    }
+  }
+
+  editHeader(evt) {
+    const id = evt.target.id;
+
+    const value = evt.target.value;
+
+    if (value) {
+      const newHeaderRow = this.state.headerRow.slice();
+
+      newHeaderRow[id] = value;
+      this.setState({headerRow: newHeaderRow});
+    } else {
+      throw Error('null value');
```

Comment:
I see you need this else statement because you can't pivot the data when the headers are null values. But it causes a weird experience where you can't delete the last character in the input box.

<img width="518" alt="screen shot 2018-04-16 at 12 29 18 am" src="https://user-images.githubusercontent.com/8146241/38790179-40b9703a-410d-11e8-9596-f2f8e8fb4cbb.png">

I think we can reuse your create new validation logic to fix this:

```
  editHeader(evt) {
    const id = evt.target.id;
    const value = evt.target.value;
    const newHeaderRow = this.state.headerRow.slice();

    newHeaderRow[id] = value;

    this.setState({headerRow: newHeaderRow});
  }
```

```
  saveEditedHeaders() {
    const newDataSet = this.state.data.slice();

    let validHeaders = true;

    this.state.headerRow.forEach(header => {
      if (header === null || header === '') {
        validHeaders = false;
      }
    });
    if (validHeaders === false) {
      this.setState({validationMessage: 'Headers cannot be blank.'});
    } else {
      newDataSet[0] = this.state.headerRow;
      this.setState({data: newDataSet, showEditHeaderForm: false,
        validationMessage: null});
    }
  }
```

In fact it seems it will be the same function as addHeaders so we could just use the same function in both cases and delete one. What do you think?